### PR TITLE
[C10D] Remove watchKey functionality from the Store.

### PR DESCRIPTION
**Pull request by kumpera**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #105674
* #105672
* #105184
* #105164
* #105163
* __->__ #105014

The feature was never fully finished and never got any adoption but
TCPStore pays the cost of twice the number of tcp connections anyway.

While the cost of all those idle connections is minimal is doesn't come for free:

- It increases the likelyhood of a connection refused failure during the initialization stampede.
- TCPStore uses poll for checking for socket availability which scales linearly on the number of sockets regardless of their status.

### CODE REVIEW

File: test/cpp/c10d/TCPStoreTest.cpp [hunk 1]
```diff
@@ -253,94 +169,6 @@ TEST(TCPStoreTest, testWatchKeyCallbackWithPrefix) {
   testWatchKeyCallback("testPrefix");
```
Comment (H-Huang) [inline test/cpp/c10d/TCPStoreTest.cpp:169]: This test and the test above also need to be removed.

Comment (kumpera) [reply test/cpp/c10d/TCPStoreTest.cpp:169]: fixed

**Review by fduwjj — APPROVED** [review]
Thanks for cleaning this up. Does this mean we previously had twice as much as sockets? That's a lot.

**Pull request closed without merging**